btree/btree.c

```diff
@@ -3529,21 +3559,429 @@ int64_t btree_iter_tick(struct m0_sm_op *smop)
+			return P_COOKIE;
+		else
+			return P_SETUP;
+	case P_COOKIE:{
+		void *cookie_segaddr = bop->bo_rec.r_key.k_cookie.segaddr;
+		if (cookie_is_valid(tree, &bop->bo_rec.r_key.k_cookie) &&
+		    !cookie_underflow_is_possible(cookie_segaddr))
+			return P_LOCK;
+		else
+			return P_SETUP;
+	}
+	case P_SETUP: {
+		bop->bo_arbor->t_height = tree->t_height;
+		bop->bo_i = level_alloc(bop->bo_arbor->t_height);
+		if (bop->bo_i == NULL)
+			return fail(bop, M0_ERR(-ENOMEM));
+		bop->bo_i->i_key_found = false;
+		return P_LOCKALL;
+	}
+	case P_LOCKALL:
+		if (bop->bo_flags & BOF_LOCKALL) {
+			//return m0_sm_op_sub(&bop->bo_op, P_LOCK, P_DOWN);
+			m0_rwlock_write_lock(&tree->t_lock);
+		}
+		/* Fall through to the next stage */
+	case P_DOWN:
+		oi->i_used = 0;
+		/* Load root node. */
+		return node_get(&oi->i_nop, tree, &tree->t_root->n_addr,
+				P_NEXTDOWN);
+	case P_NEXTDOWN:
+		if (oi->i_nop.no_op.o_sm.sm_rc == 0) {
+			struct slot    node_slot = {};
+			struct segaddr child_node_addr;
+
+			lev->l_node = oi->i_nop.no_node;
+			node_slot.s_node = oi->i_nop.no_node;
+			lev->l_seq = lev->l_node->n_seq;
+			/* verify node footer */
+			/*int rc = node_verify(lev->l_node);
+			if (rc)
+			{
+				node_op_fini(&oi->i_nop);
+				return fail(bop, rc);
+			}*/
+			oi->i_nop.no_node = NULL;
+
+			oi->i_key_found = node_find(&node_slot,
+						    &bop->bo_rec.r_key);
+			lev->l_idx = node_slot.s_idx;
+
+			if (node_level(node_slot.s_node) > 0) {
+				if (oi->i_key_found) {
+					lev->l_idx++;
+					node_slot.s_idx++;
+				}
+				node_child(&node_slot, &child_node_addr);
+
+				if (!address_in_segment(child_node_addr)) {
+					node_op_fini(&oi->i_nop);
+					return fail(bop, M0_ERR(-EFAULT));
+				}
+				oi->i_used++;
+				return node_get(&oi->i_nop, tree,
+						&child_node_addr, P_NEXTDOWN);
+			} else {
+				if (!oi->i_key_found)
+					return P_LOCK;
+				/**
+				 * If root is an internal node and it contains
+				 * only two record, if any of the record is
+				 * going to be deleted, load the other child of
+				 * root.
+				 */
+				if (oi->i_used > 0 &&
+				    node_count_rec(oi->i_level[0].l_node) == 2)
+					return handle_root_case(bop);
+
+				return P_LOCK;
+			}
+		} else {
+			node_op_fini(&oi->i_nop);
+			return fail(bop, oi->i_nop.no_op.o_sm.sm_rc);
+		}
+	case P_STORE_CHILD: {
+		/* store child of the root. */
+		oi->i_level[1].l_sib = oi->i_nop.no_node;
+		oi->i_level[1].l_sib_seq = oi->i_nop.no_node->n_seq;
+		/* Fall through to the next step */
+	}
+	case P_LOCK:
+		if (!locked(tree))
+			return lock_op_init(&bop->bo_op, &bop->bo_i->i_nop,
+					    bop->bo_arbor->t_desc, P_CHECK);
+		else
+			return P_CHECK;
+	case P_CHECK: {
+		oi->i_trial++;
+		if (!path_check(oi, tree, &bop->bo_rec.r_key.k_cookie) ||
+		    !child_node_check(oi)) {
+			if (oi->i_trial == MAX_TRIALS) {
+				if (bop->bo_flags & BOF_LOCKALL)
+					return fail(bop, -ETOOMANYREFS);
+				else
+					bop->bo_flags |= BOF_LOCKALL;
```

Comment:
sfter setting flag as BOF_LOCKALL it will check if height has changed or not and accordingly it will go to P_INIT, P_DOWN,
as the code below line:3902 is in if part not else if